Fix/70 locale a11y

adamwoodnz opened the pull request:
Fixes #70 

Branches off the initial commits in #71 with styles to get the autocomplete as close as possible to the designs and the current switcher on prod

### Screenshots

#### Desktop
| Focus | Cleared | Search |
|-|-|-|
| ![Screen Shot 2023-04-27 at 1 01 47 PM](https://user-images.githubusercontent.com/1017872/234733623-357f578b-0592-48d9-97f3-0c8ba366360f.jpg) | ![Screen Shot 2023-04-27 at 1 02 05 PM](https://user-images.githubusercontent.com/1017872/234733653-f80d7000-43d8-4559-9984-edf00d56ad6f.jpg) | ![Screen Shot 2023-04-27 at 1 02 21 PM](https://user-images.githubusercontent.com/1017872/234733672-ec0ade15-fba1-495b-9043-832d99c223c8.jpg) |

#### Mobile (emulated)
| Focus | Cleared | Search |
|-|-|-|
| ![Screen Shot 2023-04-27 at 1 05 36 PM](https://user-images.githubusercontent.com/1017872/234733929-582d1823-d11c-45f8-b7c0-d1a3e3d4100d.jpg) | ![Screen Shot 2023-04-27 at 1 05 49 PM](https://user-images.githubusercontent.com/1017872/234733943-865c97f9-5184-4b8c-8dd4-da5b6522a957.jpg) | ![Screen Shot 2023-04-27 at 1 06 02 PM](https://user-images.githubusercontent.com/1017872/234733956-8e1aeb56-fc8c-4f0d-9ed9-aa3d6aa20c45.jpg) |

#### Mobile

The autoselected input text isn't a very nice experience on my Chrome for Android, as the contextual options are displayed, but I guess that is expected, and enables the input to be cleared with one tap of delete:

<img src="https://user-images.githubusercontent.com/1017872/234741379-5bed5abc-99fe-415a-a962-414fef3f8e3b.jpg" width="400" />

iandunn committed:
wip form basics working, needs lots of style work

iandunn committed:
load autocomplete styles before theme so can override

adamwoodnz committed:
Style AlphaGov accessible autocomplete to match designs

adamwoodnz committed:
Hide native autocomplete select

adamwoodnz commented on wp-content/mu-plugins/assets/locale-switcher.js:
```diff
@@ -16,26 +16,105 @@
+					// make sure escaped
+
+
+				placeholder: 'Search languages...',
```
String needs translation. From current UI: 

<img src="https://user-images.githubusercontent.com/1017872/234743732-50a4fddf-fc8f-41a8-b1c9-937f4a4f8034.jpg" width="300" />

adamwoodnz replied:
Ah maybe it should use the string here https://github.com/WordPress/wp20.wordpress.net/pull/93/files#diff-1368bd0d70766b64532fd35e145bb6e6cea9b47f3297d37027f10d2f8865e763R30

adamwoodnz replied:
Done in https://github.com/WordPress/wp20.wordpress.net/pull/93/commits/54d4216fdcbb7da69b8b38e14e36b53239afe91c

iandunn replied:
Yeah, that looks like a good approach 👍🏻

adamwoodnz committed:
Reduce the height of the autocomplete input when displayed at the top of the list

adamwoodnz commented on wp-content/themes/twentyseventeen-wp20/header.php:
```diff
@@ -27,7 +27,8 @@
 							<?php get_template_part( 'template-parts/header/navigation', 'top' ); ?>
 							<?php Locales\locale_switcher(); ?>
-							<div class="navigation-top-menu-container" data-placeholder="<?php esc_attr_e( 'Search languages...', 'wp20' ); ?>"></div>
+							<!-- data-placeholder="<?php esc_attr_e( 'Search languages...', 'wp20' ); ?>" -->
```
Seems like this isn't required with the `label .screen-reader-text` and [placeholder in JS](https://github.com/WordPress/wp20.wordpress.net/pull/93/files#diff-c86a86024c4de465cc9785b050a285506c9b64f456cc4c7b923c48345cfc021eR85)

adamwoodnz committed:
Enable translation of locale switcher placeholder

adamwoodnz committed:
Removed unused .navigation-top-menu-container

adamwoodnz committed:
Stop locale switcher from overlapping navigation

iandunn committed:
remove WIP comments

iandunn committed:
start to translate autocomplete strings

iandunn committed:
wip started using wp-i18n

iandunn committed:
Merge branch 'trunk' into fix/70-locale-a11y

iandunn committed:
translate in js

iandunn committed:
remove debugging code

iandunn committed:
remove old select2 code

iandunn committed:
get more js translations working

iandunn committed:
switch to aliased functions

iandunn committed:
finish translations

iandunn committed:
add escaping notes

iandunn committed:
pass i18n functions into IIFE

iandunn committed:
add zindex to show input/ul when clicking

when on desktop and tab into the autocomplete it displays fine, but when _clicking_ into it the `input` and `ul` are hidden.
not sure if this will be needed after other changes though

iandunn commented on wp-content/themes/twentyseventeen-wp20/style.css:
```diff
@@ -696,132 +699,94 @@ body:not(.title-tagline-hidden) .site-branding-text {
+		display: flex;
+		flex-direction: column;
+		align-items: flex-end;
+		z-index: 1;
```
I added this in 385a511 because the input/ul was hidden on desktops when clicking into it:

<img width="707" alt="Screenshot 2023-05-03 at 6 32 05 AM" src="https://user-images.githubusercontent.com/484068/235931041-30f70194-4578-46fe-a147-5b7425e6e3b0.png">

i'm not sure if it'll be needed after other changes though

iandunn committed:
remove span because its read by VoiceOver

iandunn commented on wp-content/themes/twentyseventeen-wp20/style.css:
```diff
@@ -696,132 +699,94 @@ body:not(.title-tagline-hidden) .site-branding-text {
-
-.wp20-locale-switcher-container {
+.wp20-locale-switcher-container form {
 	position: absolute;
```
This makes it so that `.site-content-contain` is covered by the dropdown on mobile:

<img width="362" alt="Screenshot 2023-05-03 at 7 44 56 AM" src="https://user-images.githubusercontent.com/484068/235951491-0115203e-fcbe-4c49-b444-d9b64ecefe63.png">

On `trunk` it pushes the content down though:

<img width="409" alt="Screenshot 2023-05-03 at 7 47 46 AM" src="https://user-images.githubusercontent.com/484068/235952231-67ecfae2-73f4-43f9-94d6-b2a1900726a5.png">

I don't see a good way to fix that. One problem I kept running into with these PRs is that the menu DOM structure is different than the locale list structure. The menu toggle button is a separate element from the menu itself, while the switcher "toggle" and dropdown are all contained in a single element.

Another part might be that [Accessible Autocomplete doesn't add classes for the current state](https://github.com/alphagov/accessible-autocomplete/issues/384). The `input` can be styles by matching `aria-expanded=true`, though; and the `ul` with `.autocomplete__menu--visible`.

adamwoodnz replied:
I think this is fixed now. I've added a wrapper to the nav menu so that it behaves more like the locale one. Try it out.

iandunn commented on wp-content/themes/twentyseventeen-wp20/style.css:
```diff
@@ -502,13 +502,16 @@ body:not(.title-tagline-hidden) .site-branding-text {
 }
 
+.navigation-top .wrap {
+	display: flex;
```
This works well for the desktop, but I noticed this on mobile:

<img width="557" alt="Screenshot 2023-05-03 at 8 22 17 AM" src="https://user-images.githubusercontent.com/484068/235962242-b6a91098-b0fd-4df1-b79c-d5db779a1e32.png">

I haven't come up with a good fix.

adamwoodnz replied:
Should also be fixed by the new wrapper.

iandunn replied:
Good idea, looks great 👍🏻

iandunn commented on wp-content/themes/twentyseventeen-wp20/style.css:
```diff
@@ -696,132 +699,94 @@ body:not(.title-tagline-hidden) .site-branding-text {
 }
 
-.wp20-locale-switcher-container .select2-container--open .select2-selection--single .select2-selection__rendered {
+.autocomplete__hint {
```
I'm not sure why, but the first word isn't capitalized in the hint, even though it is in the raw data and in the dropdown:

<img width="455" alt="Screenshot 2023-05-03 at 8 25 19 AM" src="https://user-images.githubusercontent.com/484068/235963079-4dd248ec-8b1d-45c7-9ed9-ec24e3e00a93.png">

It's odd because the text doesn't seem to show up in the DOM, and it's not set via CSS `content` either 🤔 I checked the Autocomplete source and it wasn't obvious there.

I don't think it's related to any of the styles, and it was happening in #71 too.

`text-transform: capitalize` works in some cases, but would be wrong in ones like `Español De Colombia` where the `De` should be lowercase.

We are using the hint in an unintended way in order to match the design, but I'm not sure why that'd be related 🤷🏻‍♂️

adamwoodnz replied:
I only get lowercase when I've entered lowercase in the input.

Looks like it might be this line, where the input text replaces the start of the option https://github.com/alphagov/accessible-autocomplete/blob/main/src/autocomplete.js#L452

`hintValue` is bound here https://github.com/alphagov/accessible-autocomplete/blob/main/src/autocomplete.js#L490

iandunn replied:
Ah, good find. Yeah, `jQuery( '.autocomplete__hint' ).val()` shows the value, but FF/Chrome inspectors don't 🤷🏻‍♂️ 

Doesn't look like anything we can change easily, so I'd say it's just a quirk, and not worry about it 👍🏻 

If we really have to I guess we could set up a mutation observer to overwrite it w/ the correct capitalization, but that may have a11y consequences, etc. I'm sure there's some reason they're doing the `substr` bit. I guess if we have to do that, it'd be better to just add our own custom element for sighted users, and leave the hint hidden for screen readers. The observer could copy the hint value into the custom element.

iandunn reviewed (commented):
I cleaned up some of my WIP JS, added translations, etc. I'm running into a few CSS issues that I haven't found solutions for, though.

adamwoodnz committed:
Fix layout issues

- Push content down when menus are open
- Fix flex layout of nav menu on mobile

iandunn committed:
add border to search autocomplete bottom on desktop

mobile doesn't need it b/c of the border from `#masthead`

iandunn reviewed (approved):
Great work, this is looking good in my tests 👍🏻
